enhancements/user-monitoring.md

```diff
@@ -0,0 +1,248 @@
+
+Out of scope are:
+- Definition of Dashboards.
+- Deployment of Prometheus instances in user namespaces.
+- Deployment of Alertmanager instances user namespaces.
+
+## Proposal
+
+![](user-monitoring-deployment.png)
+
+The existing cluster monitoring stack is going to be extended with additional Prometheus (P) servers
+and one additional Prometheus Operator (PO) reconciling user namespaces
+as well as the newly added Prometheus servers.
+
+The cluster Prometheus Operator (PO) continues to operate the cluster-level Prometheus.
+We will continue to use a centralized Alertmanager (A) cluster.
+
+Querying data from a single, multi-tenant interface is done using the Thanos Querier (TQ) component.
+Soft tenancy is continued to be ensured by the existing topology leveraging kube-rbac-proxy [1]
+and prom-label-proxy [2].
+
+Differences between cluster- and user-level is mainly upgrade guarantees.
+As we keep cluster-level Prometheus immutable, we can guarantee a 100% cluster upgrade success
+but that rate may be lower for any additional Prometheus server scraping user namespaces.
+There are edge cases where an automated upgrade is not possible in the face of user monitoring.
+If users declare i.e. queries in ServiceMonitors which break in newer Prometheus versions
+we most likely cannot perform an automated migration, as PromQL is too complex to achieve that.
+
+Instead the "upgradable" condition in the ClusterOperator will be set to false.
+Routing configurations for alerts to be sent to specific teams need to be communicated/configured by an infrastructure owner.
+We will look into better mechanisms after GA.
+With any OpenShift deployment, we only install the cluster-level part
+and allow infrastructure owner to "enable" the addition of user monitoring.
+
+### User Stories
+
+
+#### US1
+
+As a service owner, I’d like to configure a ServiceMonitor CRD so that Red Hat’s supported monitoring solution is able to scrape metrics for my service.
+
+#### US2
+
+As a service owner, I’d like to configure alerting rules for my service so that Red Hat’s supported monitoring solution notifies me if problems turn up.
+
+#### US3
+
+As a member of the operations team, I want to configure Prometheus one time so that all existing and any new namespaces will have access.
+
+#### US4
+
+As a member of the operations team, I'd like to add my own exporters so that I can monitor infrastructure services that are not monitored out-of-the-box.
+
+Note: This covers adding ServiceMonitor and PrometheusRule CRDs.
+
+#### US5
+
+As a developer working on a Red Hat product not part of the infrastructure, I’d like to integrate my metrics into a single, supported Prometheus stack so that they are available for my customers.
+
+#### US6
+
+As a service owner, I’d like to query the metrics I have access to from the OpenShift console.
+
+#### US7
+
+As a member of the operations team, I’d like to make sure that one team doesn't have access to metrics from another team.
```

Comment:
Within one cluster? Within one Prometheus sets, right?

Reply:
i imagine as a developer/user i'd also like to make sure other teams/tenants don't have access to my metrics.
(same use case, different stakeholder)

Reply:
I think replacing team with tenant here sets the expectations appropriately. How teams are distributed in an OpenShift cluster is subjective to each customer.

Reply:
replaced team with tenant.